### PR TITLE
Fix build cpp demo

### PR DESCRIPTION
**Pull request by gaulouis**
Thanks you to consider my "contribution".

@see https://github.com/antlr/antlr4/issues/1457 (namespace pollution)
@see https://github.com/antlr/antlr4/issues/1464 (warnings polution)
@see https://github.com/antlr/antlr4/issues/1443 (build Linux error)

### CODE REVIEW

File: runtime/Cpp/demo/TLexer.g4 [hunk 1]
```diff
@@ -66,7 +66,7 @@ OpenCurly: '{' -> pushMode(Mode1);
 CloseCurly: '}' -> popMode;
 QuestionMark: '?';
 Comma: ',' -> skip;
-Dollar: '$' -> more, mode(Mode1), type(DUMMY);
```
Comment (KvanTTT) [inline runtime/Cpp/demo/TLexer.g4]: I guess the tokens section
```ANTLR
tokens {
	DUMMY	
}
```
can be also removed for now. It's not used anywhere.

Comment (gaulouis) [inline runtime/Cpp/demo/TLexer.g4]: Thanks,

I will remove it...

**Comment by parrt** [comment]
@mike-lischke do you bless this?

**Comment by gaulouis** [comment]
"do you bless this?"

No !

1] options{ superClass} bugged
```
ParserHeader(parser, funcs, atn, sempredFuncs, superClass = {Parser}) ::= <<
<namedActions.context>

class <parser.name> : public antlr4::<superClass> {
```
need to be replaced by 
```
ParserHeader(parser, funcs, atn, sempredFuncs, superClass = {antlr4::Parser}) ::= <<
```

2] Lexer namespace pollution not clean

3] antlr4 namespace present in impelementation file
I need to add: ``LexerRuleContextHeader() ::= "antlr4::RuleContext"`` according to ``LexerRuleContext() ::= "RuleContext"``

**Comment by parrt** [comment]
Ok, so I should close or let you keep working on this branch/PR?

**Comment by gaulouis** [comment]
I would like to familiarize myself with git.

Please leave it open so I can have a full review proccess.

PS: These are just small modifications.

**Comment by gaulouis** [comment]
I have break some thing. I'am sorry. How can i fix it ?

File: tool/resources/org/antlr/v4/tool/templates/codegen/Cpp/Cpp.stg [hunk 2]
```diff
@@ -263,10 +263,10 @@ bool <if (r.factory.g.lexer)><lexer.name><else><parser.name><endif>::<r.name>Sem
 
 //--------------------------------------------------------------------------------------------------
 
-ParserHeader(parser, funcs, atn, sempredFuncs, superClass = {Parser}) ::= <<
+ParserHeader(parser, funcs, atn, sempredFuncs, superClass = {antlr4::Parser}) ::= <<
 <namedActions.context>
 
-class <parser.name> : public antlr4::<superClass> {
+class <parser.name> : public <superClass> {
```
Comment (mike-lischke) [inline tool/resources/org/antlr/v4/tool/templates/codegen/Cpp/Cpp.stg]: That will not work well. A superclass doesn't necessarily have to be from the antlr4 namespace, but could be an own class.

**Comment by mike-lischke** [comment]
PLEASE don't mix many unrelated changes into a single PR. It's very difficult to review them and almost impossible to accept or reject if something is wrong. Make only related changes into a PR and rather issue several PRs then.

Some notes (though I'd like to make them on individual PRs):

1. Don't change the C++ template lightheaded. The problem with multiple semicolons is known and there is no solution currently. The culprit here is that sometimes templates are used in other templates in various nesting levels. Sometimes a semicolon is provided by a higher level, sometimes it is not.
2. The demo: the grammar is more than just a demonstration. It contains everything to make ANTLR generate all necessary constructs, so that we can see if the code generation works properly. Don't remove elements from the grammar! It's not primarily thought as a demonstration for the end user (the demo app is however). There is currently no other way to test completeness of code generation.
3. Bug #1443 is only a missing update of the linux demo app. I simply haven't come around to do that.
4. Some template changes are plain wrong (e.g. the antlr4:: prefix for <SuperClass>).

So, I recommend to close this PR and make the changes instead into individual PRs for individual reviews.

**Comment by gaulouis** [comment]
Thank you for your valuable informations.

**Comment by KvanTTT** [comment]
@mike-lischke I agree with you: the tokens section should be existitng, but the `type` command should be used in another case.

**Comment by gaulouis** [comment]
I have do:
```
-Dollar: '$' -> more, mode(Mode1), type(DUMMY);

+Dollar: '$' -> more, mode(Mode1);
+Ampersand: '&' -> type(DUMMY);
```
However, I have no rules that use Dollar or Ampersand. Do i need to create a rule ?

**Comment by mike-lischke** [comment]
@gaulouis No, it's just fine like that. It doesn't count that we don't use a rule, the code generation still kicks in for every rule we have in a grammar. Your change is fine, just check that the demo app still parses the example input, so that new users have an easy start.

**Pull request closed without merging**